src/diagram/workflow_builder.rs

```diff
@@ -0,0 +1,547 @@
+
+use super::{
+    fork_clone::DynForkClone, impls::DefaultImpl, split_chain, transform::transform_output,
+    BuiltinTarget, Diagram, DiagramError, DiagramOperation, DiagramScope, DynInputSlot, DynOutput,
+    NextOperation, NodeOp, NodeRegistry, OperationId, SourceOperation, SplitOpParams,
+};
+
+struct Vertex<'a> {
+    op_id: &'a OperationId,
+    op: &'a DiagramOperation,
+    in_edges: Vec<usize>,
+    out_edges: Vec<usize>,
+}
+
+struct Edge<'a> {
+    source: SourceOperation,
+    target: &'a NextOperation,
+    state: EdgeState<'a>,
+}
+
+enum EdgeState<'a> {
+    Ready {
+        output: DynOutput,
+        /// The node that initially produces the output, may be `None` if there is no origin.
+        /// e.g. The entry point, or if the output passes through a `join` operation which
+        /// results in multiple origins.
+        origin: Option<&'a NodeOp>,
+    },
+    Pending,
+}
+
+pub(super) fn create_workflow<'a, Streams: StreamPack>(
+    scope: DiagramScope<Streams>,
+    builder: &mut Builder,
+    registry: &NodeRegistry,
+    diagram: &'a Diagram,
+) -> Result<(), DiagramError> {
+    // first create all the vertices
+    let mut vertices: HashMap<&OperationId, Vertex> = diagram
+        .ops
+        .iter()
+        .map(|(op_id, op)| {
+            (
+                op_id,
+                Vertex {
+                    op_id,
+                    op,
+                    in_edges: Vec::new(),
+                    out_edges: Vec::new(),
+                },
+            )
+        })
+        .collect();
+
+    // init with some capacity to reduce resizing. HashMap for faster removal.
+    let mut edges: HashMap<usize, Edge> = HashMap::with_capacity(diagram.ops.len() * 2);
+
+    // process start separately because we need to consume the scope input
+    match &diagram.start {
+        NextOperation::Builtin { builtin } => match builtin {
+            BuiltinTarget::Terminate => {
+                // such a workflow is equivalent to an no-op.
+                builder.connect(scope.input, scope.terminate);
+                return Ok(());
+            }
+            BuiltinTarget::Dispose => {
+                // bevy_impulse will immediate stop with an `CancellationCause::Unreachable` error
+                // if trying to run such a workflow.
+                return Ok(());
+            }
+        },
+        NextOperation::Target(op_id) => {
+            edges.insert(
+                edges.len(),
+                Edge {
+                    source: SourceOperation::Builtin {
+                        builtin: super::BuiltinSource::Start,
+                    },
+                    target: &diagram.start,
+                    state: EdgeState::Ready {
+                        output: scope.input.into(),
+                        origin: None,
+                    },
+                },
+            );
+            vertices.get_mut(&op_id).unwrap().in_edges.push(0);
```

Comment:
We should either provide a rationale for why this unwrap cannot panic or this case should be covered as a `DiagramError`.